src/goto-symex/README.md

```diff
@@ -50,44 +45,85 @@ The \ref bmct creates a goto_symext to symbolically execute the
 the parameter "`new_symbol_table`" passed as a reference to the various
 functions in goto_symext.
 
-The main symbolic execution loop code is goto_symext::symex_step(). This
+The main symbolic execution loop code is \ref goto_symext::symex_step. This
 function case-switches over the type of the instruction that we're
 currently executing, and calls various other functions appropriate to
 the instruction type, i.e. goto_symext::symex_function_call() if the
 current instruction is a function call, goto_symext::symex_goto() if the
 current instruction is a goto, etc.
 
+\subsection Loop and recursion unwinding
+
+Each backwards goto and recursive call has a separate counter
+(handled by \ref cbmc or another driver program, see the `–unwind` and
+`–unwind-set` options). The symbolic execution includes constant
+folding so loops that have a constant / bounded number of iterations will often
+be handled completely (assuming the unwinding limit is sufficient).
+When an unwind or recursion limit is reached, an assertion can be added to
+explicitly show when analysis is incomplete.
+
+\subsection \ref goto_symext::clean_expr
+
+Before any expression is incorporated into the output equation set it is passed
+through \ref goto_symext::clean_expr and thus \ref goto_symext::dereference,
+whose primary purpose is to remove dereference operations. It achieves this
+using the \ref value_sett stored in \ref goto_symex_statet, replacing `*x` with
+a construct such as
+`x == &candidate1 ? candidate1 : x == &candidate2 ? candidate2 : x$object`
+
+Note the `x$object` fallback candidate, which is known as a *failed symbol* or
+*failed object*, and represents the unknown object pointed to by `x` when
+neither of the candidates (`candidate1` and `candidate2` here) matched as
+expected. This is of course unsound, since `x$object` and `y$object` are always
+distinct, even if `x` and `y` might actually alias, but at least it ensures
+writes to and subsequent reads from `x` are related.
+
+Despite its name, the \ref goto_symext::dereference function is also responsible
+for calling \ref goto_symex_statet::rename, which is responsible for both SSA
+renaming symbols and for applying constant propagation where possible.
+
 \subsection symex-path Path exploration
 
 By default, CBMC symbolically executes the entire program, building up
 an equation representing all instructions, which it then passes to the
 solver. Notably, it _merges_ paths that diverge due to a goto
 instruction, forming a constraint that represents having taken either of
-the two paths; the code for doing this is goto_symext::merge_goto().
-
-CBMC can operate in a different mode when the `--paths` flag is passed
-on the command line. This disables path merging; instead, CBMC executes
-a single path at a time, calling the solver with the equation
+the two paths; the code for doing this is \ref goto_symext::merge_goto.
+
+Symex can operate in a different mode when the `--paths` flag is passed
```

Comment:
Goto-symex (to avoid confusion with the 'symex' that has been moved out into a separate repo)